### DISCUSSION TITLE
Saving an item with 2 primary key causes one to override the other

### DISCUSSION THREAD

**Discussion by yannisoo** [mikro-orm/mikro-orm#5215] [General]
### Describe the bug

Hi, I just started using MikroORM. I have a very strange issue when saving an item in database for the first time. I'm using a 2 primary keys as follow, one for joining tables the other as a public uiid.
```typescript
  @PrimaryKey({ type: 'uuid' })
  uuid = randomUUID().toString();

  @PrimaryKey({ type: 'bigint', autoincrement: true })
  @SerializedPrimaryKey()
  id!: number;
```

However when saving my item into database for the first time the auto incremental id overrides the uuid and makes them both the regular id. Thanks in advance
```typescript
    console.log(user);
    await this.em.flush();
    console.log(user);
```
![image](https://github.com/mikro-orm/mikro-orm/assets/36623256/ee5df428-2084-42b6-b104-97f4ff7ea189)

### Reproduction

see screenshots 

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.1.0

### Node.js version

20.11

### Operating system

macOS (nest is ran through a docker container)

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
Drop the `@SerializedPrimaryKey()` decorator, that's for mongo driver, and it's not supposed to be used in combination with `@PrimaryKey()` (this is the part that is probably causing your issue).

But then there is the bigger issue - having a composite key with db-generated values is not supported, you need to provide the value for all the keys, this is a limitation of the current implementation (at least it used to be, and I don't recall fixing this). Do you really need to have a composite PK here? If so, do you need both parts of it to be generated and already unique on their own?

>  I'm using a 2 primary keys as follow, one for joining tables the other as a public uiid.

This sounds like you just need a unique hash, and you will be fine with just the autoincrement PK.

**Comment by nicomouss** [comment]
@yannisoo I am also using a combo of serial Id plus uuid on my entities. The way you want to deal with that is by having this setup:

```ts
@PrimaryKey({ type: 'bigint', autoincrement: true })
id: number;

@Property({ type: 'uuid' })
uuid: string = randomUUID().toString();
```

The database (postgresql) will treat the uuid column as a column with a `unique` constraint by default, as it is of type `uuid`. So you don't need to include that column as part of the primary key.

Besides that, it is not optimal to use uuids as primary keys in relational databases. It leads to bigger clustered index (compared to just using a sequential integer). Also it leads to a lot of clustered index pages re-organization when inserting new rows, because uuids are not sequential (although if you are using uuidv7 instead of uuidv4, then it's better in that regard, but then another problem is that a timestamp information is leaking in the uuidv7 which might be a security issue for your entities).

Hence the best of both world is to just use a sequential id as the primary key (which is basically an internal technical/system id for your entities), and a separate uuid (that is the public id for your entities).

Hope that helps.
